bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/scanner/ResponseQueueReader.java

```diff
@@ -66,16 +80,25 @@ public synchronized Row getNextMergedRow() throws IOException {
       }
     }
 
-    Preconditions.checkState(
-        builder == null,
-        "End of stream marker encountered while merging a row.");
-    Preconditions.checkState(
-        lastResponseProcessed,
-        "Should only exit merge loop with by returning a complete Row or hitting end of stream.");
+    Preconditions.checkState(builder == null,
+      "End of stream marker encountered while merging a row.");
+    Preconditions.checkState(lastResponseProcessed,
+      "Should only exit merge loop with by returning a complete Row or hitting end of stream.");
     return null;
   }
 
   private ResultQueueEntry<ReadRowsResponse> getNext() throws IOException {
+    int unprocessedRequestCount = outstandingRequestCount.get() + resultQueue.size();
+
+    // If there are currently less than or equal to half of the capacityCap, then ask gRPC to
+    // request more results in a batch. Batch requests are more efficient that reading one at
+    // a time.
+    while (!completionMarkerFound.get()
+        && unprocessedRequestCount <= capacityCap - batchRequestSize) {
+      call.request(batchRequestSize);
```

Comment:
if completionMarkerFound, it means that the server indicated that there is no more data.  There's no need to read more.